### ISSUE TITLE
Problem while decrypting signature on some video

### ISSUE THREAD

**Issue by davinkevin** [axet/vget#36]
Hi, 

I report an error I find after trying to download a new video at the following url "https://www.youtube.com/watch?v=RzpZ2w3f04U&feature=youtube_gdata"

I have the following stacktrace : 
```
Exception in thread "Thread-11" java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.charAt(String.java:646)
	at com.github.axet.vget.vhs.YouTubeParser$DecryptSignature.s(YouTubeParser.java:91)
	at com.github.axet.vget.vhs.YouTubeParser$DecryptSignature.decrypt(YouTubeParser.java:119)
	at com.github.axet.vget.vhs.YouTubeParser.extractUrlEncodedVideos(YouTubeParser.java:643)
	at com.github.axet.vget.vhs.YouTubeParser.extractHtmlInfo(YouTubeParser.java:496)
	at com.github.axet.vget.vhs.YouTubeParser.streamCpature(YouTubeParser.java:249)
	at com.github.axet.vget.vhs.YouTubeParser.extractLinks(YouTubeParser.java:203)
	at com.github.axet.vget.vhs.YouTubeParser.extract(YouTubeParser.java:663)
	at com.github.axet.vget.info.VGetParser.info(VGetParser.java:16)
	at com.github.axet.vget.VGet.extract(VGet.java:293)
	at lan.dk.podcastserver.manager.worker.downloader.YoutubeDownloader.download(YoutubeDownloader.java:109)
	at lan.dk.podcastserver.manager.worker.downloader.AbstractDownloader.startDownload(AbstractDownloader.java:65)
	at lan.dk.podcastserver.manager.worker.downloader.AbstractDownloader.run(AbstractDownloader.java:56)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:201)
	at com.sun.proxy.$Proxy118.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
```
I have the following value in the debug mode when the error rises up.
![image](https://cloud.githubusercontent.com/assets/1970922/7405278/d90b02c6-eef0-11e4-94ab-2cb4cc5c06ba.png)

**Comment by ahmadftx** [comment]
i got the same error

**Comment by TigranTsat** [comment]
Same for me on http://www.youtube.com/watch?v=vGZdqgmM2FQ

**Comment by davinkevin** [comment]
Hello, same thing on other video : https://www.youtube.com/watch?v=o-xmhFMZsGU

![image](https://cloud.githubusercontent.com/assets/1970922/16713166/007c740a-469f-11e6-8f58-c40c2064e517.png)

Value from this breakpoint : 
- detailMessage : 
```
<eval>:17:54 Expected , but found ;
Jr=function(a){return Ir(function(a,c){return Hr(a,c)};var sr={rj:function(a){a.reverse()},
                                                      ^
```
- decodeScript : 
```
tr=function(a){a=a.split("");sr.p9(a,3);sr.rj(a,16);sr.Pk(a,58);sr.Pk(a,34);sr.rj(a,33);return a.join("")};
ur=function(a,b){this.b=a||0;this.spatialAudioType=b||0};
xr=function(a,b,c,d,e,f,k,l){this.id=""+a;this.g=0<=b.indexOf("/mp4")?1:0<=b.indexOf("/webm")?2:0<=b.indexOf("/x-flv")?3:0<=b.indexOf("/vtt")?4:0;this.mimeType=b;this.b=k||0;this.A=l||0;this.audio=c||null;this.video=d||null;this.Rb=e||null;this.Te=f||null;this.o=vr[wr(this)]||""};
wr=function(a){return a.id.split(";",1)[0]};
yr=function(a){return"9"==a.o||"("==a.o};
zr=function(a){return!(a.audio&&a.video)};
Ar=function(a){return 0<=a.indexOf("opus")||0<=a.indexOf("vorbis")||0<=a.indexOf("mp4a")};
Br=function(a){return 0<=a.indexOf("vp9")||0<=a.indexOf("vp8")||0<=a.indexOf("avc1")};
Cr=function(a,b,c){this.name=a;this.id=b;this.isDefault=c};
Fr=function(a,b,c,d,e,f,k,l,m,n){this.width=a;this.height=b;if(!f)a:{f=Math.max(a,b);a=Math.min(a,b);b=Dr[0];for(var t=0;t<Dr.length;t++){var u=Dr[t],x=Er[u];if(f>=Math.floor(16*x/9)*Fr.prototype.g||a>=x*Fr.prototype.g){f=b;break a}b=u}f="tiny"}this.quality=f;this.pE=Er[this.quality];this.isAccelerated=!!k;this.fps=c||0;this.A=e||0;this.projectionType=d||0;(c=l)||(c=this.fps,d=Er[this.quality],c=0==d?O("YTP_AUTO"):d+"p"+(55<c?"60":49<c?"50":39<c?"48":""));this.qualityLabel=c;this.o=m||"";this.b=!1!==
n};
Gr=function(a,b){return new Cd(function(c,d){b.$a=function(a){Qj(a)?c(a):d(a)};
b.onError=d;b.bc=d;Vj(a,b)})};
Hr=function(a,b){function c(e,f,k){return Od(e,function(e){return 0>=f||403===e.status?Hd(e):d(k).then(function(){return c(Gr(a,b),f-1,Math.pow(2,4-f+1)*k)})})}
function d(a){return new Cd(function(b){(0,window.setTimeout)(b,a)})}
return c(Gr(a,b),3,1E3)};
Jr=function(a){return Ir(function(a,c){return Hr(a,c)};var sr={rj:function(a){a.reverse()},
p9:function(a,b){a.splice(0,b)},
Pk:function(a,b){var c=a[0];a[0]=a[b%a.length];a[b]=c}};
```

- playerScript : https://gist.github.com/davinkevin/00c1f7557e71fd7839613bf3dd16ec6b
- sig : `83C83C1DD6D31B1C6E9BEB6C438D3E3BF656C50E921.D010FDCD965A50256AFC0F66FD991C09F5A14477`

@axet : I've had this problem of download to this issue, because the title I've choose was coherent with my new current problem, but don't you think we should find a better way to inform you about a download problem ? I think an issue by video can be helpful, so you can prioritize the fix to be made in function of type of issue (Signature Error, Https error or other problem...). The Label of issue are really helpful too for that...

Thanks for your help, and if you need more information / help, I'll be glad to hear it

**Comment by axet** [comment]
@davinkevin this is well know problem. To solve it we need to understand how youtube encryption works or well port 'youtube.dl' python command line client to java. I'm not ready to do any of this. You can go deep into it or keep ignoring exception and wait.

**Comment by davinkevin** [comment]
I already investigate on this, it appears the script found by the `extractDecodeFunctions` contains syntax error. 

Original result of extractDecodeFunctions : 
```
tr=function(a){a=a.split("");sr.p9(a,3);sr.rj(a,16);sr.Pk(a,58);sr.Pk(a,34);sr.rj(a,33);return a.join("")};
ur=function(a,b){this.b=a||0;this.spatialAudioType=b||0};
xr=function(a,b,c,d,e,f,k,l){this.id=""+a;this.g=0<=b.indexOf("/mp4")?1:0<=b.indexOf("/webm")?2:0<=b.indexOf("/x-flv")?3:0<=b.indexOf("/vtt")?4:0;this.mimeType=b;this.b=k||0;this.A=l||0;this.audio=c||null;this.video=d||null;this.Rb=e||null;this.Te=f||null;this.o=vr[wr(this)]||""};
wr=function(a){return a.id.split(";",1)[0]};
yr=function(a){return"9"==a.o||"("==a.o};
zr=function(a){return!(a.audio&&a.video)};
Ar=function(a){return 0<=a.indexOf("opus")||0<=a.indexOf("vorbis")||0<=a.indexOf("mp4a")};
Br=function(a){return 0<=a.indexOf("vp9")||0<=a.indexOf("vp8")||0<=a.indexOf("avc1")};
Cr=function(a,b,c){this.name=a;this.id=b;this.isDefault=c};
Fr=function(a,b,c,d,e,f,k,l,m,n){this.width=a;this.height=b;if(!f)a:{f=Math.max(a,b);a=Math.min(a,b);b=Dr[0];for(var t=0;t<Dr.length;t++){var u=Dr[t],x=Er[u];if(f>=Math.floor(16*x/9)*Fr.prototype.g||a>=x*Fr.prototype.g){f=b;break a}b=u}f="tiny"}this.quality=f;this.pE=Er[this.quality];this.isAccelerated=!!k;this.fps=c||0;this.A=e||0;this.projectionType=d||0;(c=l)||(c=this.fps,d=Er[this.quality],c=0==d?O("YTP_AUTO"):d+"p"+(55<c?"60":49<c?"50":39<c?"48":""));this.qualityLabel=c;this.o=m||"";this.b=!1!==
n};
Gr=function(a,b){return new Cd(function(c,d){b.$a=function(a){Qj(a)?c(a):d(a)};
b.onError=d;b.bc=d;Vj(a,b)})};
Hr=function(a,b){function c(e,f,k){return Od(e,function(e){return 0>=f||403===e.status?Hd(e):d(k).then(function(){return c(Gr(a,b),f-1,Math.pow(2,4-f+1)*k)})})}
function d(a){return new Cd(function(b){(0,window.setTimeout)(b,a)})}
return c(Gr(a,b),3,1E3)};
Jr=function(a){return Ir(function(a,c){return Hr(a,c)};var sr={rj:function(a){a.reverse()},
p9:function(a,b){a.splice(0,b)},
Pk:function(a,b){var c=a[0];a[0]=a[b%a.length];a[b]=c}};
```

The element I put instead of it : 
```js
tr=function(a){a=a.split(\"\");sr.p9(a,3);sr.rj(a,16);sr.Pk(a,58);sr.Pk(a,34);sr.rj(a,33);return a.join(\"\")};\nur=function(a,b){this.b=a||0;this.spatialAudioType=b||0};\nxr=function(a,b,c,d,e,f,k,l){this.id=\"\"+a;this.g=0<=b.indexOf(\"/mp4\")?1:0<=b.indexOf(\"/webm\")?2:0<=b.indexOf(\"/x-flv\")?3:0<=b.indexOf(\"/vtt\")?4:0;this.mimeType=b;this.b=k||0;this.A=l||0;this.audio=c||null;this.video=d||null;this.Rb=e||null;this.Te=f||null;this.o=vr[wr(this)]||\"\"};\nwr=function(a){return a.id.split(\";\",1)[0]};\nyr=function(a){return\"9\"==a.o||\"(\"==a.o};\nzr=function(a){return!(a.audio&&a.video)};\nAr=function(a){return 0<=a.indexOf(\"opus\")||0<=a.indexOf(\"vorbis\")||0<=a.indexOf(\"mp4a\")};\nBr=function(a){return 0<=a.indexOf(\"vp9\")||0<=a.indexOf(\"vp8\")||0<=a.indexOf(\"avc1\")};\nCr=function(a,b,c){this.name=a;this.id=b;this.isDefault=c};\nFr=function(a,b,c,d,e,f,k,l,m,n){this.width=a;this.height=b;if(!f)a:{f=Math.max(a,b);a=Math.min(a,b);b=Dr[0];for(var t=0;t<Dr.length;t++){var u=Dr[t],x=Er[u];if(f>=Math.floor(16*x/9)*Fr.prototype.g||a>=x*Fr.prototype.g){f=b;break a}b=u}f=\"tiny\"}this.quality=f;this.pE=Er[this.quality];this.isAccelerated=!!k;this.fps=c||0;this.A=e||0;this.projectionType=d||0;(c=l)||(c=this.fps,d=Er[this.quality],c=0==d?O(\"YTP_AUTO\"):d+\"p\"+(55<c?\"60\":49<c?\"50\":39<c?\"48\":\"\"));this.qualityLabel=c;this.o=m||\"\";this.b=!1!==\nn};\nGr=function(a,b){return new Cd(function(c,d){b.$a=function(a){Qj(a)?c(a):d(a)};\nb.onError=d;b.bc=d;Vj(a,b)})};\nHr=function(a,b){function c(e,f,k){return Od(e,function(e){return 0>=f||403===e.status?Hd(e):d(k).then(function(){return c(Gr(a,b),f-1,Math.pow(2,4-f+1)*k)})})}\nfunction d(a){return new Cd(function(b){(0,window.setTimeout)(b,a)})}\nreturn c(Gr(a,b),3,1E3)};\nJr=function(a){return Ir(function(a,c){return Hr(a,c)})};var sr={rj:function(a){a.reverse()},\np9:function(a,b){a.splice(0,b)},\nPk:function(a,b){var c=a[0];a[0]=a[b%a.length];a[b]=c}};
```

There is just a missing `)}`. 
After hard replacing it, the code seems to work well 😄 

Check this on https://github.com/davinkevin/vget/blob/%2336-signature-deconding-issue/src/main/java/com/github/axet/vget/vhs/YouTubeParser.java#L262 

![image](https://cloud.githubusercontent.com/assets/1970922/16713355/713579d4-46a5-11e6-9617-e4c5aee09524.png)

**Comment by davinkevin** [comment]
Solved with released 1.1.34 (which is the 100th release 🎂  !!) 

@axet : Last time, let me know if you're investigating the problem, that will allow me to not make a PR for nothing... 😉 

@ahmadftx and @TigranTsat : I've tested with the link you submitted, seems to work for me, do you confirm ?
